crypto/pem/pem_info.c

```diff
@@ -127,74 +118,48 @@ STACK_OF(X509_INFO) *PEM_X509_INFO_read_bio_ex(BIO *bp, STACK_OF(X509_INFO) *sk,
                 goto start;
             }
             pp = &(xi->crl);
-        } else if (strcmp(name, PEM_STRING_RSA) == 0) {
-            d2i = (D2I_OF(void)) d2i_RSAPrivateKey;
-            if (xi->x_pkey != NULL) {
-                if (!sk_X509_INFO_push(ret, xi))
-                    goto err;
-                if ((xi = X509_INFO_new()) == NULL)
-                    goto err;
-                goto start;
-            }
-
-            xi->enc_data = NULL;
-            xi->enc_len = 0;
-
-            xi->x_pkey = X509_PKEY_new();
-            if (xi->x_pkey == NULL)
-                goto err;
-            ptype = EVP_PKEY_RSA;
-            pp = &xi->x_pkey->dec_pkey;
-            if ((int)strlen(header) > 10) /* assume encrypted */
-                raw = 1;
-        } else
+        } else if (strcmp(name, PEM_STRING_RSA) == 0
 #ifndef OPENSSL_NO_DSA
-        if (strcmp(name, PEM_STRING_DSA) == 0) {
-            d2i = (D2I_OF(void)) d2i_DSAPrivateKey;
-            if (xi->x_pkey != NULL) {
-                if (!sk_X509_INFO_push(ret, xi))
-                    goto err;
-                if ((xi = X509_INFO_new()) == NULL)
-                    goto err;
-                goto start;
-            }
-
-            xi->enc_data = NULL;
-            xi->enc_len = 0;
-
-            xi->x_pkey = X509_PKEY_new();
-            if (xi->x_pkey == NULL)
-                goto err;
-            ptype = EVP_PKEY_DSA;
-            pp = &xi->x_pkey->dec_pkey;
-            if ((int)strlen(header) > 10) /* assume encrypted */
-                raw = 1;
-        } else
+                   || strcmp(name, PEM_STRING_DSA) == 0
 #endif
 #ifndef OPENSSL_NO_EC
-        if (strcmp(name, PEM_STRING_ECPRIVATEKEY) == 0) {
-            d2i = (D2I_OF(void)) d2i_ECPrivateKey;
+                   || strcmp(name, PEM_STRING_ECPRIVATEKEY) == 0
+#endif
+                   || strcmp(name, PEM_STRING_PKCS8INF) == 0
+                   || strcmp(name, PEM_STRING_PKCS8) == 0) {
             if (xi->x_pkey != NULL) {
                 if (!sk_X509_INFO_push(ret, xi))
                     goto err;
                 if ((xi = X509_INFO_new()) == NULL)
                     goto err;
                 goto start;
             }
-
+            if (strcmp(name, PEM_STRING_RSA) == 0) {
+                d2i = (D2I_OF(void)) d2i_RSAPrivateKey;
```

Comment:
Instead of using these type specific functions that are deprecated it might be a better idea to call OSSL_DECODER directly.

Reply:
Good thought - I'll simply use `d2i_AutoPrivateKey()` in all "*PRIVATE KEY" cases.

Reply:
This will make it possible to enclose RSA key in BEGIN EC PRIVATE KEY - not sure it is that big deal though.

Reply:
Calling OSSL_DECODER directly allows specifying expected type.

Reply:
Yeah, but using OSSL_DECODER directly appears more cumbersome, and I don't think it's worth checking for specific and possibly non-matching key type indicators preceding the "PRIVATE KEY" tag text.

Reply:
> I don't think it's worth checking for specific and possibly non-matching key type indicators preceding the "PRIVATE KEY" tag text

This sounds worrisome.  Are you saying that if something says "BEGIN EC PRIVATE KEY" in the PEM marker, and you're expecting an RSA key, you'll try the ASN1 decode anyway?  Or do I misunderstand what you're saying?  (Most likely that)

Reply:
You actually understand it right. OTOH the above sounds more worrisome than it is.

AFAICS, in former times PEM encodings included specific texts such as 
"{RSA,DSA,EC} PRIVATE KEY", but meanwhile "BEGIN PRIVATE KEY" is most common. In any case the latter is sufficient because what actually counts in the end is the ASN.1 encoded stuff, which has its own tagging, and the PEM-level tags are meanwhile just used as separators.

So if I expect an RSA key and I get one, I'm happy regardless whether the PEM marker said RSA or whatever. And if I get an EC key although the marker wrongly supported my expectation of an RSA key, that EC key simply won't match when used because the type will be wrong (and for instance not match the type in an allegedly corresponding cert). This is not (much) worse than if I confused that RSA key with some other RSA key, which the PEM parser would not have any chance to catch.

Reply:
After thinking a little more about it, I've found and implemented a simple way of generically matching any key type name given at PEM level,
which also leads to more efficient parsing since other types are then not tried at ASN.1 level.